src/utils/format_json/open_telemetry.c

```diff
@@ -0,0 +1,278 @@
+
+  CHECK(json_add_string(g, "key"));
+  CHECK(json_add_string(g, label.name));
+
+  CHECK(json_add_string(g, "value"));
+  CHECK(yajl_gen_map_open(g)); /* BEGIN AnyValue */
+  CHECK(json_add_string(g, "stringValue"));
+  CHECK(json_add_string(g, label.value));
+  CHECK(yajl_gen_map_close(g)); /* END AnyValue */
+
+  CHECK(yajl_gen_map_close(g)); /* END KeyValue */
+  return 0;
+}
+
+static int number_data_point(yajl_gen g, metric_t const *m) {
+  CHECK(yajl_gen_map_open(g)); /* BEGIN NumberDataPoint */
+
+  CHECK(json_add_string(g, "attributes"));
+  for (size_t i = 0; i < m->label.num; i++) {
+    CHECK(key_value(g, m->label.ptr[i]));
+  }
+  CHECK(yajl_gen_array_close(g));
+
+  CHECK(json_add_string(g, "timeUnixNano"));
+  CHECK(yajl_gen_integer(g, CDTIME_T_TO_NS(m->time)));
+
+  switch (m->family->type) {
+  case METRIC_TYPE_COUNTER:
+    CHECK(json_add_string(g, "asInt"));
+    CHECK(yajl_gen_integer(g, m->value.counter));
+    break;
+  case METRIC_TYPE_GAUGE:
+    CHECK(json_add_string(g, "asDouble"));
+    CHECK(yajl_gen_integer(g, m->value.gauge));
+    break;
+  case METRIC_TYPE_UNTYPED:
+    // TODO
+    assert(0);
+  }
+
+  CHECK(yajl_gen_map_close(g)); /* END NumberDataPoint */
+  return 0;
+}
+
+static int gauge(yajl_gen g, metric_family_t const *fam) {
+  CHECK(yajl_gen_map_open(g)); /* BEGIN Gauge */
+
+  CHECK(json_add_string(g, "dataPoints"));
+  CHECK(yajl_gen_array_open(g));
+  for (size_t i = 0; i < fam->metric.num; i++) {
+    CHECK(number_data_point(g, fam->metric.ptr + i));
+  }
+  CHECK(yajl_gen_array_close(g));
+
+  CHECK(yajl_gen_map_close(g)); /* END Gauge */
+  return 0;
+}
+
+static int sum(yajl_gen g, metric_family_t const *fam) {
+  CHECK(yajl_gen_map_open(g)); /* BEGIN Sum */
+
+  CHECK(json_add_string(g, "dataPoints"));
+  CHECK(yajl_gen_array_open(g));
+  for (size_t i = 0; i < fam->metric.num; i++) {
+    CHECK(number_data_point(g, fam->metric.ptr + i));
+  }
+  CHECK(yajl_gen_array_close(g));
+
+  char const *aggregation_temporality_cumulative = "2";
+  CHECK(json_add_string(g, "aggregationTemporality"));
+  CHECK(json_add_string(g, aggregation_temporality_cumulative));
+
+  CHECK(json_add_string(g, "isMonotonic"));
+  CHECK(yajl_gen_bool(g, true));
+
+  CHECK(yajl_gen_map_close(g)); /* END Sum */
+  return 0;
+}
+
+static int metric(yajl_gen g, metric_family_t const *fam) {
+  CHECK(yajl_gen_map_open(g)); /* BEGIN Metric */
+
+  CHECK(json_add_string(g, "name"));
+  CHECK(json_add_string(g, fam->name));
+
+  // TODO(octo): populate the "unit" field.
+  // CHECK(json_add_string(g, "unit"));
+  // CHECK(json_add_string(g, "1"));
+
+  if (fam->help != NULL) {
+    CHECK(json_add_string(g, "description"));
+    CHECK(json_add_string(g, fam->help));
+  }
+
+  switch (fam->type) {
+  case METRIC_TYPE_COUNTER:
+    CHECK(json_add_string(g, "sum"));
+    CHECK(sum(g, fam));
+    break;
+  case METRIC_TYPE_GAUGE:
+    CHECK(json_add_string(g, "gauge"));
+    CHECK(gauge(g, fam));
```

Comment:
`metric()` caller does it with `CHECK()`, so same error could be logged up to 6 times.

Reply:
Fixed in 1bd6d8c7220fe9f00e9155f4a8dc4570b200ba08